distributed/scheduler.py

```diff
@@ -2467,14 +2506,18 @@ def validate_waiting(self, key):
 
     def validate_processing(self, key):
         ts = self.tasks[key]
-        assert not ts.waiting_on
+        # speculative doesn't update waiting_on, so ensure that all dependencies are
+        # processing on the same worker if not `waiting_on`
+        assert (
+            not ts.waiting_on or len({dts.processing_on for dts in ts.waiting_on}) == 1
+        )
         ws = ts.processing_on
         assert ws
         assert ts in ws.processing
         assert not ts.who_has
         for dts in ts.dependencies:
-            assert dts.who_has
-            assert ts in dts.waiters
+            assert dts.who_has or ts.speculative
+            assert ts in dts.waiters or ts.speculative
```

Comment:
I would expect there to be a separate `validate_speculative` method that was applied to tasks in the `speculative` state rather than reusing this method.